### PR TITLE
Implement labelledExamples

### LINKED ISSUE #163: Port of QuickCheck's labelledExamples?
It doesn't look like haskell-hedgehog even has this!  I would love to debug my generators and labelling with it.

I'm actually so keen that I might give a shot at implementing it, but I wouldn't how to do it and I can't even steal from haskell-hedgehog here...  Any advice/tips/mentoring?

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes #163

This introduces some duplication but, more generally, I'm not sure it's doing exactly the right thing.
Particularly: how is it meant to behave on a failing property?  What about None results?  Also, should it only
print the journal logs, as it's doing, or also the result logs?  And did I make all the correct modifications in
my "takeSmallest" copy?  Lots of questions.

I'd gladly have this land in a "we're not totally sure"/experimental feature and dogfood it in a project of
mine, which might give some better insights on what design choices are better.  (But I can also copy this impl
into my project, so I wouldn't be blocked either - pure FP ftw.)

### CODE REVIEW

**Comment by charleso** [comment]
@dwijnand I'm definitely ok with taking some experimental code. I _think_ my only preference here would not be to duplicate any of that logic though (those 2 functions in particular are critical to how it all hangs together). If we can either share what we can with common functions and possibly have some parameters/callbacks that might avoid us having to keep the two in sync. In particular, I know those functions have been refactored for `Future` support.

https://github.com/hedgehogqa/scala-hedgehog/pull/147

I don't have any specific suggestions though I'm afraid.

**Comment by dwijnand** [comment]
Alright, I'll seek to dedup.

**Comment by dwijnand** [comment]
Can this be merged and released?  I think I left this in a state I was very happy with: useful and deduped - but happy to have another look if a fresh review spots anything.

**Comment by eed3si9n** [comment]
Is there a way of doing partest neg equivalent to see what gets printed on the screen when some test fails?

**Comment by dwijnand** [comment]
No, there's no setup for testing the behaviour of runner, just unit tests for the core's behaviour.  (See also test/README.md 🙂)

**Comment by eed3si9n** [comment]
My interest is in the negative behavior like in this case how the example affect the test failure output, not so much the details of different runners.

**Comment by dwijnand** [comment]
Before:

```
[info] + hedgehog.LabelledExamplesTest.testReverse: OK, passed 100 tests
[info] > 72% nonempty
[info] > 28% empty
[info] - hedgehog.LabelledExamplesTest.testReverseFail: Falsified after 20 passed tests
[info] > List(a, b)
[info] > === Not Equal ===
[info] > --- lhs ---
[info] > List(b, a)
[info] > --- rhs ---
[info] > List(a, b)
[info] > 65% empty
[info] > 35% nonempty
```

After:

```
[info] + hedgehog.LabelledExamplesTest.testReverse: OK, passed 100 tests
[info] > 69% nonempty List(a)
[info] > 31% empty List()
[info] - hedgehog.LabelledExamplesTest.testReverseFail: Falsified after 19 passed tests
[info] > List(a, b)
[info] > === Not Equal ===
[info] > --- lhs ---
[info] > List(b, a)
[info] > --- rhs ---
[info] > List(a, b)
[info] > 78% empty List()
[info] > 21% nonempty List(a)
```

using

```scala
    , property("testReverse", testReverse).withExamples
    , property("testReverseFail", testReverseFail).withExamples

...

  def testReverse: Property =
    for {
      xs <- Gen.alpha.list(Range.linear(0, 10)).forAll
        .classify("empty", _.isEmpty)
        .classify("nonempty", _.nonEmpty)
    } yield xs.reverse.reverse ==== xs

  def testReverseFail: Property =
    for {
      xs <- Gen.alpha.list(Range.linear(0, 10)).forAll
        .classify("empty", _.isEmpty)
        .classify("nonempty", _.nonEmpty)
    } yield xs.reverse ==== xs
```

**Comment by Kevin-Lee** [comment]
Looks good! @dwijnand Would you mind updating docs as well?

**Comment by charleso** [comment]
👍 Looks good @dwijnand. You should definitely feel comfortable merging if and someone else are happy with it. Don't wait for me.

Just one final (probably unhelpful) thought was whether it would just be easier to always collect the examples on every run which might simply that code. It might not make sense, or not worth spending on more time on it, and in any case feel free to ignore me. :)

Thanks again!

**Comment by Kevin-Lee** [comment]
That minimum coverage with `cover` works really well so the test failed. 😆

```
[info] - hedgehog.examples.ReverseTest.reverse: Falsified after 100 passed tests
[info] > Insufficient coverage.
[info] > 95% nonempty 50% ✓ List(a)
[info] > 5% empty 50% ✗ List()
```

**Comment by Kevin-Lee** [comment]
@charleso I think it's good to merge it now and we can consider your thought afterwards. 
Dale created this PR about a month ago and it's a nice feature to have so we should probably not spend more time on this.

**Comment by Kevin-Lee** [comment]
@dwijnand As Charles said, please feel free to merge it. If you want to merge it but can't do it, please let me know.

**Comment by Kevin-Lee** [comment]
v0.5.0 has been released.
https://search.maven.org/search?q=g:qa.hedgehog%20AND%20a:hedgehog-*%20AND%20v:0.5.0

File: core/shared/src/main/scala/hedgehog/core/PropertyT.scala [hunk 1]
```diff
@@ -247,6 +269,15 @@ case class DiscardCount(value: Int) {
     DiscardCount(value + 1)
 }
 
+/** Whether the report should include an example for each label. */
+sealed trait WithExamples
```
Comment (charleso) [inline core/shared/src/main/scala/hedgehog/core/PropertyT.scala:273]: No longer used? :)

Comment (dwijnand) [reply core/shared/src/main/scala/hedgehog/core/PropertyT.scala:273]: Where did that come from?  #175

**Pull request merged**